File: backend/models/Application.js
import mongoose from 'mongoose';

// ─── Application Model ──────────────────────────────
// Represents a candidate's application to a job posting.
// Tracks application status through the hiring pipeline.

const applicationSchema = new mongoose.Schema({
  // The job being applied to
  job: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Job',
    required: [true, 'Application must reference a job'],
  },
  // The candidate who applied
  applicant: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'Application must have an applicant'],
  },
  // Current status in the hiring pipeline
  status: {
    type: String,
    enum: ['applied', 'shortlisted', 'interview', 'offered', 'rejected', 'withdrawn'],
    default: 'applied',
  },
  // Cover letter or additional message from the candidate
  coverLetter: {
    type: String,
    default: '',
  },
  // Resume URL at the time of application
  resumeUrl: {
    type: String,
    default: null,
  },
  // Recruiter notes about this application
  notes: {
    type: String,
    default: '',
  },
  // Assessment score if candidate took an assessment for this job
  assessmentScore: {
    type: Number,
    default: null,
  },
}, {
  timestamps: true,
});

// Prevent duplicate applications (one candidate per job)
applicationSchema.index({ job: 1, applicant: 1 }, { unique: true });
applicationSchema.index({ applicant: 1, status: 1 });
applicationSchema.index({ job: 1, status: 1 });

const Application = mongoose.model('Application', applicationSchema);
export default Application;

File: backend-spring/src/main/java/com/learn2hire/model/Application.java
package com.learn2hire.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.CompoundIndexes;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "applications")
@CompoundIndexes({
    @CompoundIndex(name = "job_applicant_unique", def = "{'job': 1, 'applicant': 1}", unique = true),
    @CompoundIndex(name = "applicant_status", def = "{'applicant': 1, 'status': 1}"),
    @CompoundIndex(name = "job_status", def = "{'job': 1, 'status': 1}")
})
public class Application {

    @Id
    @JsonProperty("_id")
    private String id;

    private String job; // Job ID reference

    private String applicant; // User ID reference

    @Builder.Default
    private String status = "applied"; // applied, shortlisted, interview, offered, rejected, withdrawn

    @Builder.Default
    private String coverLetter = "";

    private String resumeUrl;

    @Builder.Default
    private String notes = "";

    private Double assessmentScore;

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;
}
